### ISSUE TITLE
[llvm18] Is this current behavior correct?

### ISSUE THREAD

**Issue by erickq** [llvm/llvm-project#65430]
```
bool correctness{true};
struct A {
    bool operator==(A const&oth) const {
	return true;
    }
};

struct B {
    bool operator==(B const&oth) const {
	return false;
    }
};

struct C {
    bool operator==(C const&oth) const {
	correctness=false;
	return false;
    }
};

typedef std::tuple<A,B,C> tuple_t ;

int test_it (void) {
    tuple_t x,y;
    return ( !(x==y) ) && correctness ;
}
```

This test case does not report an error in GCC11 and later versions. The tuple class overloads the operator==() method. Different structs also overload the operator==() method. Due to the introduction of the patch, the llvm18 reports an error when compiling the test case. Is this reasonable?

GCC11:https://godbolt.org/z/qfjxeGjjj
GCC10.5:https://godbolt.org/z/WzdjEec18

**Comment by erickq** [comment]
It seems that the modification of this patch (https://reviews.llvm.org/D155387) is introduced.

**Comment by erickq** [comment]
@shafik

**Comment by shafik** [comment]
I don't see any diagnostic using clang trunk: https://godbolt.org/z/s8d4jqbv3

Can you provide a godbolt link that has clang trunk producing a diagnostic.

**Comment by erickq** [comment]
> I don't see any diagnostic using clang trunk: https://godbolt.org/z/s8d4jqbv3
> 
> Can you provide a godbolt link that has clang trunk producing a diagnostic.

LLVM17 does not report an error, but LLVM18 reports an error. Godbolt doesn't seem to support llvm18 yet.

When I build llvm18 (the latest in the community), an error is reported. You can try it locally, thank you.

![image](https://github.com/llvm/llvm-project/assets/12108571/cd60dd8c-e68c-45e9-8ed9-326e05b9471e)

**Comment by erickq** [comment]
> > I don't see any diagnostic using clang trunk: https://godbolt.org/z/s8d4jqbv3
> > Can you provide a godbolt link that has clang trunk producing a diagnostic.
> 
> LLVM17 does not report an error, but LLVM18 reports an error. Godbolt doesn't seem to support llvm18 yet.
> 
> When I build llvm18 (the latest in the community), an error is reported. You can try it locally, thank you.
> 
> ![image](https://user-images.githubusercontent.com/12108571/265877040-cd60dd8c-e68c-45e9-8ed9-326e05b9471e.png)

@shafik

**Comment by shafik** [comment]
So clang trunk was built today: https://github.com/compiler-explorer/compiler-workflows/blob/main/build-status.md

and in general clang trunk should be up to date.

Can you do a `--version` and see what hash you have.

**Comment by erickq** [comment]
> So clang trunk was built today: https://github.com/compiler-explorer/compiler-workflows/blob/main/build-status.md
> 
> and in general clang trunk should be up to date.
> 
> Can you do a `--version` and see what hash you have.

![image](https://github.com/llvm/llvm-project/assets/12108571/8c1dc78a-2a11-4678-981e-2bc8b845e877)

**Comment by erickq** [comment]
![image](https://github.com/llvm/llvm-project/assets/12108571/33b332d7-4a02-4604-86c4-e89cd8d6ca42)

I just tried the latest upstream llvm (fd4c80dec929d0dc00d92745ffb0f62b3ff32dd8), and this problem can be reproduced.

**Comment by erickq** [comment]
@shafik

**Comment by erickq** [comment]
![image](https://github.com/llvm/llvm-project/assets/12108571/7fe76e11-e18c-4763-b875-3c75ae3ab091)

https://godbolt.org/z/4eMooEf4c

It's strange that I'm aligning the local commit with the godbolt commit. However, an error is reported locally, but not on Godbolt.

@shafik

**Comment by erickq** [comment]
I think I know why. thanks
